cpp/src/arrow/util/compression_zlib.cc

```diff
@@ -246,7 +246,7 @@ class GZipCompressor : public Compressor {
     //  again with the same value of the flush parameter and more output space
     //  (updated avail_out), until the flush is complete (deflate returns
     //  with non-zero avail_out)."
-    return FlushResult{bytes_written, (bytes_written == 0)};
+    return FlushResult{bytes_written, (stream_.avail_out == 0)};
```

Comment:
Should we also check for `Z_BUF_ERROR` here?

Reply:
Done.